geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/iterators/CartesianProductIterable.scala

```diff
@@ -14,32 +14,51 @@ package org.locationtech.geomesa.utils.iterators
  * way to query the list size that is independent of the iterator itself.
  * (That is, asking for the size does not exhaust any iterator.)
  *
+ * NB:  The first sequence is the least significant; that is, it will
+ * increment fast while the last sequence is the most significant (will
+ * increment slowly).
+ *
  * @param seqs the list-of-lists whose items are to be recombined
  * @tparam T the type of items
  */
-case class CartesianProductIterable[T](seqs:Seq[Seq[T]]) extends Iterable[Seq[T]] {
-  // how many (total) combinations there will be in an iterator
-  lazy val expectedSize : Long = seqs.map(seq => seq.size.toLong).product
 
-  // create an iterator that will visit all possible combinations
-  // (every time you call this, you get a NEW iterator)
-  def iterator : Iterator[Seq[T]] = new Iterator[Seq[T]] {
-    // variable index that counts off the combinations
-    var index : Long = 0L
-    // current value (combination)
-    private def value : Seq[T] =
-      seqs.foldLeft(Seq[T](),index)((t,src) => t match { case (seqSoFar,idx) => {
-        val modulus : Int = src.size
-        val itemIndex : Int = (idx % modulus).toInt
-        val item : T = src(itemIndex)
-        (seqSoFar++Seq(item), idx / modulus)
-      }})._1
-    def next() : Seq[T] = {
-      val result = if (hasNext) value else null
-      index = index + 1L
+case class CartesianProductIterable(seqs: Seq[Seq[_]]) extends Iterable[Seq[_]] {
+  lazy val expectedSize: Int = seqs.map(_.size).product
+
+  def iterator: Iterator[Seq[_]] = new Iterator[Seq[_]] {
+    val n = seqs.size
+    val maxes = seqs.map(seq => seq.size)
+    val indexes = scala.collection.mutable.MutableList.fill(seqs.size)(0)
+    var nextItem: Option[Seq[_]] = if (isValid) Option(realize) else None
```

Comment:
If we are using vars for optimization, we can do null checks rather than instantiating Options.  @cne1x can you try without Options and see if it is any faster/slower?

Also, I wonder if using a Vector would be better than/different from using a MutableList....

Reply:
For fun, removing `Option` reduced the new runtime by an additional ~5%; introducing `ArraySeq` and `Vector` reduced the new runtime by an additional ~65%.  I had forgotten we talked about `Vector` yesterday at our Scala for Cats session; thanks for the reminder!

Reply:
any thoughts on comparisons between `Vector` and `Array` or `ArrayBuffer`? Those are usually my go tos, but I haven't compared them.

Reply:
![image](https://user-images.githubusercontent.com/1149270/47936378-d0e54a00-deb3-11e8-9c99-a472630e6925.png)

Reply:
Yeah, `Array` with `ArrayBuffer` have indistinguishable performance (at least in this case) from `Vector` and `ArraySeq`.

Reply:
Pushed a commit with the refined version.